### PR TITLE
[JS] cwe-327 (Weak or vulnerable cryptography usage) added

### PR DESCRIPTION
**Pull request by monkey-junkie**
Hi team!
I've wrote detection of usage of weak block cipher algorithms like block ciphers in ECB mode (known as chosen plaintext attack) and for common crypto flaw - static IV for block ciphers. 
Either of such flaws could led to sensitive data leak.

More information:

- OWASP - https://owasp.org/www-community/vulnerabilities/Using_a_broken_or_risky_cryptographic_algorithm

- OWASP cheatsheet - https://cheatsheetseries.owasp.org/cheatsheets/Cryptographic_Storage_Cheat_Sheet.html

- Real world example - https://www.ambionics.io/blog/prestashop-privilege-escalation

- CPA - https://en.wikipedia.org/wiki/Chosen-plaintext_attack

### CODE REVIEW

**Comment by intrigus-lgtm** [comment]
Is this similar to https://github.com/github/codeql/blob/master/javascript/ql/src/Security/CWE-327/BrokenCryptoAlgorithm.qhelp and https://github.com/github/codeql/blob/master/javascript/ql/src/Security/CWE-327/BrokenCryptoAlgorithm.ql ?

**Comment by monkey-junkie** [comment]
> Is this similar to https://github.com/github/codeql/blob/master/javascript/ql/src/Security/CWE-327/BrokenCryptoAlgorithm.qhelp and https://github.com/github/codeql/blob/master/javascript/ql/src/Security/CWE-327/BrokenCryptoAlgorithm.ql ?

Hi!
Not exactly. Proposed query could be used for detection of incorrect usage of secure algorithms (for example, AES-128 ECB which could be vulnerable for CPA), when mentioned query is about known-weak algorithms. 
Maybe, it should be more appropriate to merge proposed query with https://github.com/github/codeql/blob/master/javascript/ql/src/Security/CWE-327/BrokenCryptoAlgorithm.ql

File: javascript/ql/src/experimental/Security/CWE-327/BrokenSymmetricCrypto.ql [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+class ConstIVSink extends BrokenSymmetricCryptoSink {
+  ConstIVSink() {
+    exists(MethodCallNode call, BasicBlock createCihperBlock, BasicBlock ivBlock, string alg |
+      call = moduleImport("crypto").getAMethodCall("createCipheriv") and
+      createCihperBlock = call.(InvokeNode).getArgument(2).getBasicBlock() and
+      ivBlock = call.(InvokeNode).getArgument(2).getALocalSource().getBasicBlock() and
+      alg = call.(InvokeNode).getArgument(0).toString()
+    |
+      createCihperBlock.length() != ivBlock.length() and
```
[the remark this answers is not in the document]
Comment (monkey-junkie) [reply javascript/ql/src/experimental/Security/CWE-327/BrokenSymmetricCrypto.ql:47]: Hi!
Yes, exactly, OFB and CTR with the static IV are vulnerable to CPA. I tried to use .legnth() method of BasicBlock to distinguish blocks with IV initialization and its actual usage. But now I can see, that I was wrong and I should use getContainer instead.
Thanks!

**Comment by intrigus-lgtm** [comment]
@monkey-junkie In case you didn't know:
> ^ Pro tip: in the "Files changed" view, you have the option of adding review suggestions to a batch and commit them all at once.

File: javascript/ql/src/experimental/Security/CWE-327/BrokenSymmetricCrypto.ql (hunk 1 above)
[the remark this answers is not in the document]
Comment (monkey-junkie) [reply javascript/ql/src/experimental/Security/CWE-327/BrokenSymmetricCrypto.ql]: I've updated pr according all your comments, please review :)

**Comment by erik-krogh** [comment]
> * Is there ever a valid reason to reuse a given IV for multiple plaintexts?

I think this is the mistake that Sony made in the Playstation 3 [article](https://www.engadget.com/2010-12-29-hackers-obtain-ps3-private-cryptography-key-due-to-epic-programm.html).   
I'm not sure if this was a chosen chosen-plaintext attack.

**Comment by monkey-junkie** [comment]
> After studying this a little more I start to wonder if this should be moved to `BrokenCryptoAlgorithm`. The chosen-plaintext vector seemed really interesting at first, but now it seems to me these crypto algorithms are insecure even if the attacker can't choose the plaintext.
> 
> Two questions:
> 
> * Is there ever a valid reason to use ECB when an outsider _can't_ control the plaintext?
> * Is there ever a valid reason to reuse a given IV for multiple plaintexts?
> 
> If not, I think we can move these sinks into `BrokenCryptoAlgorithm` and just flag them regardless of whether the plaintext is tainted.
> 
> If there are valid reasons (and we could thus get FPs from `BrokenCryptoAlgorithm`), I'd appreciate it if the documentation is more specific to chosen-plaintext attacks rather than the generic "incorrect usage" term. For example, the `@description` could be:
> 
> > Encrypting user-controlled plaintexts with ECB or with a reused IV enables an attacker to recover the secret key via chosen-plaintext attack.
> 
> (keep in mind I'm not a cryptographer so the above description may be wrong, in particular the part about actually recovering the key)

Hi!
> Is there ever a valid reason to use ECB when an outsider can't control the plaintext?

No.
> Is there ever a valid reason to reuse a given IV for multiple plaintexts?

According to[ RFC for CTR mode](https://tools.ietf.org/html/rfc3686#section-2.1), "The same IV and key combination MUST NOT be used more than once" so, technically, it is safe to use static IV with unique encryption key per message, but unique encryption key is very rare case.  

In sum, I think you are right and whe should move query to `BrokenCryptoAlgorithm` with additional check for key uniqueness (based on `getContainer`, like for unique IV in my `ConstIVSink ` class).
So I'm going to add my updated pr to `BrokenCryptoAlgorithm`, if you dont mind.

**Comment by asgerf** [comment]
> So I'm going to add my updated pr to BrokenCryptoAlgorithm, if you dont mind.

Thanks. Please go ahead.

**Comment by asgerf** [comment]
Closing the PR due to inactivity, but feel free to reopen if you wish to continue work on it.

Overall I think the ECB and reused-IV stuff could be useful to have in query suite in some form, so I'd be happy to see it re-opened. In retrospect I wish I had not suggested moving it into the non-experimental `BrokenCryptoAlgorithm` query because the requirements are higher for such queries, and it's just more work to integrate with the abstraction we use internally. Merging as an experimental query would have been better than closing the PR unmerged like this.

**Pull request closed without merging**